package org.gwaspi.cli;

import java.io.IOException;
import java.util.List;
import org.apache.felix.scr.annotations.Component;
import org.apache.felix.scr.annotations.Reference;
import org.gwaspi.constants.cImport.ImportFormat;
import org.gwaspi.netCDF.loader.GenotypesLoadDescription;
import org.gwaspi.netCDF.operations.GWASinOneGOParams;
import org.gwaspi.threadbox.MultiOperations;

/**
 * @author Fernando Mu√±iz Fernandez
 * IBE, Institute of Evolutionary Biology (UPF-CSIC)
 * CEXS-UPF-PRBB
 */
@Component
class LoadGenotypesScriptCommand extends AbstractScriptCommand {

	@Reference
	private MultiOperations multiOperations;

	protected void bindMultiOperations(MultiOperations multiOperations) {
		this.multiOperations = multiOperations;
	}

	protected void unbindMultiOperations(MultiOperations multiOperations) {

		if (this.multiOperations == multiOperations) {
			this.multiOperations = null;
		}
	}

	LoadGenotypesScriptCommand() {
		super("load_genotypes");
	}

	@Override
	public boolean execute(List<String> args) throws IOException {

		GWASinOneGOParams gwasParams = new GWASinOneGOParams();

		// checking study
		int studyId = prepareStudy(args.get(1), true);
		boolean studyExists = checkStudy(studyId);

		if (studyExists) {
			ImportFormat format = ImportFormat.compareTo(args.get(2));
			String newMatrixName = args.get(4);
			String description = args.get(5);

			GenotypesLoadDescription loadDescription = new GenotypesLoadDescription(
					args.get(6), // File 1
					args.get(8), // Sample Info file
					args.get(7), // File 2
					studyId, // StudyId
					format, // Format
					newMatrixName, // New Matrix name
					description, // Description
					gwasParams.getChromosome(),
					gwasParams.getStrandType(),
					gwasParams.getGtCode() // Gt code (deprecated)
					);
			multiOperations.loadMatrixDoGWASifOK(
					loadDescription, // Format
					Boolean.parseBoolean(args.get(3)), // Dummy samples
					false, // Do GWAS
					gwasParams); // gwasParams (dummy)

			return true;
		}

		return false;
	}
}
